lib/ansible/playbook/conditional.py

```diff
@@ -114,23 +114,27 @@ def _check_conditional(self, conditional, templar, all_vars):
         if isinstance(conditional, bool):
             return conditional
 
-        if C.CONDITIONAL_BARE_VARS:
-            if conditional in all_vars and VALID_VAR_REGEX.match(conditional):
-                display.deprecated('evaluating %s as a bare variable, this behaviour will go away and you might need to add |bool'
-                                   ' to the expression in the future. Also see CONDITIONAL_BARE_VARS configuration toggle.' % conditional, "2.12")
-                conditional = all_vars[conditional]
-
         if templar.is_template(conditional):
             display.warning('conditional statements should not include jinja2 '
                             'templating delimiters such as {{ }} or {%% %%}. '
                             'Found: %s' % conditional)
+
+        bare_vars_warning = False
+        if C.CONDITIONAL_BARE_VARS:
+            if conditional in all_vars and VALID_VAR_REGEX.match(conditional):
+                conditional = all_vars[conditional]
+                bare_vars_warning = True
+
         # make sure the templar is using the variables specified with this method
         templar.available_variables = all_vars
 
         try:
             # if the conditional is "unsafe", disable lookups
             disable_lookups = hasattr(conditional, '__UNSAFE__')
             conditional = templar.template(conditional, disable_lookups=disable_lookups)
+            if not isinstance(conditional, bool) and bare_vars_warning:
```

Comment:
I debate whether issuing the warning here is catching some of the edge cases.

Take the following example:

```
- debug:
  when: thing
  vars:
    thing: other
    other: false
```

This is a little contrived for brevity, but is the users intent to evaluate the string `"other"` or have it double interpolated to the value of `False`?

Reply:
@sivel Good food for thought. I haven't altered what happens for the example task above on this branch versus devel. But if brackets are used instead the bare variable warning doesn't occur.
```
- debug:
  when: thing
  vars:
    thing: "{{ other }}"
    other: false
```
In your example without brackets I'm unsure if the non-bracketed edge case has the wrong behavior. Since there is a bare var in the mix getting a warning for it seems like it could be appropriate. Thoughts?

Reply:
I feel pretty happy with this as it is. My reasoning is that if the behavior is going to stay for the user when CONDITIONAL_BARE_VARS changes then we don't need to warn them about anything.